pkg/limits/stream_metadata.go

```diff
@@ -0,0 +1,232 @@
+
+import (
+	"sync"
+)
+
+// AllFunc is a function that is called for each stream in the metadata.
+// It is used to count per tenant active streams.
+// Note: The collect function should not modify the stream metadata.
+type AllFunc = func(tenant string, partitionID int32, stream Stream)
+
+// UsageFunc is a function that is called for per tenant streams.
+// It is used to read the stream metadata for a specific tenant.
+// Note: The collect function should not modify the stream metadata.
+type UsageFunc = func(partitionID int32, stream Stream)
+
+// StreamMetadata represents the ingest limits interface for the stream metadata.
+type StreamMetadata interface {
+	// All iterates over all streams and applies the given function.
+	All(fn AllFunc)
+
+	// Usage iterates over all streams for a specific tenant and collects the overall usage,
+	// e.g. the total active streams and the total size of the streams.
+	Usage(tenant string, fn UsageFunc)
+
+	// Store updates or creates the stream metadata for a specific tenant and partition.
+	Store(tenant string, partitionID int32, streamHash, recTotalSize uint64, recordTime, bucketStart, bucketCutOff int64, evict bool)
+
+	// Evict removes all streams that have not been seen for a specific time.
+	Evict(cutoff int64) map[string]int
+
+	// EvictPartitions removes all unassigned partitions from the metadata for every tenant.
+	EvictPartitions(partitions []int32)
+}
+
+// Stream represents the metadata for a stream loaded from the kafka topic.
+// It contains the minimal information to count per tenant active streams and
+// rate limits.
+type Stream struct {
+	Hash        uint64
+	LastSeenAt  int64
+	TotalSize   uint64
+	RateBuckets []RateBucket
+}
+
+// RateBucket represents the bytes received during a specific time interval
+// It is used to calculate the rate limit for a stream.
+type RateBucket struct {
+	Timestamp int64  // start of the interval
+	Size      uint64 // bytes received during this interval
+}
+
+type stripeLock struct {
+	sync.RWMutex
+	// Padding to avoid multiple locks being on the same cache line.
+	_ [40]byte
+}
+
+type streamMetadata struct {
+	stripes []map[string]map[int32][]Stream // stripe -> tenant -> partitionID -> streamMetadata
+	locks   []stripeLock
+}
+
+func NewStreamMetadata(size int) StreamMetadata {
+	s := &streamMetadata{
+		stripes: make([]map[string]map[int32][]Stream, size),
+		locks:   make([]stripeLock, size),
+	}
+
+	for i := range s.stripes {
+		s.stripes[i] = make(map[string]map[int32][]Stream)
+	}
+
+	return s
+}
+
+func (s *streamMetadata) All(fn AllFunc) {
+	for i := range s.stripes {
+		s.locks[i].RLock()
+
+		for tenant, partitions := range s.stripes[i] {
+			for partitionID, partition := range partitions {
+				for _, stream := range partition {
+					fn(tenant, partitionID, stream)
+				}
+			}
+		}
+
+		s.locks[i].RUnlock()
+	}
+}
+
+func (s *streamMetadata) Usage(tenant string, fn UsageFunc) {
+	for i := range s.stripes {
+		s.locks[i].RLock()
```

Comment:
How is the performance if we invert the locks from per-partition to per-tenant? That would avoid having to acquire all stripes on read.